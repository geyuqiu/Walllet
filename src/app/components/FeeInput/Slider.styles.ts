export const sliderStyle = `
	input[type='range'] {
    background: linear-gradient(to right, #FBC457 0%, #FBC457 50%, #C7C9CD 50%, #C7C9CD 100%);
	  border-radius: 8px;
    width: 100%;
	  outline: none;
	  transition: background 450ms ease-in;
	  -webkit-appearance: none;
  }
  
  input[type=range]::-webkit-slider-thumb {
	  box-shadow: 1px 1px 1px #000000;
	  border: 1px solid #000000;
	  height: 16px;
	  width: 6px;
	  border-radius: 22px;
	  background: #E8FBFF;
	  cursor: pointer;
	  -webkit-appearance: none;
	  margin-top: -6.5px;
	}
	input[type=range]:hover::-webkit-slider-thumb{
	  background: #FBC457;
	}
	input[type=range]::-webkit-slider-runnable-track {
	  height: 4px;
	}
`;
