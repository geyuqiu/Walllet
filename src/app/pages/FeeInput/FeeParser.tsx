export const parse = (value: string): string => {
	let replaced = value.replace(/[^0-9.,]/g, "").replace(/,/g, ".");

	replaced = replaceGreaterThan(replaced, /\./g, 2, "");
	const split = replaced.split(".");
	if (split.length > 1 && split[1].length > 8) {
		replaced = Number(replaced).toFixed(8);
	}
	return replaced;
};

const replaceGreaterThan = (s: string, regex: RegExp, index: number, replacedWith: string): string => {
	let i = 0;
	return s.replace(regex, (match: string) => {
		i += 1;
		if (i >= index) {
			return replacedWith;
		}
		return match;
	});
};

export const hundredMillion = Math.pow(10, 8);
