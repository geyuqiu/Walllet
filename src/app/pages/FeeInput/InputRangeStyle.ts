export const inputRangeStyle = `
	input[type='range'] {
    background: linear-gradient(to right, #FBC457 0%, #FBC457 0%, #C7C9CD 0%, #C7C9CD 100%);
	  border-radius: 8px;
    width: 100%;
	  outline: none;
	  transition: background 450ms ease-in;
	  -webkit-appearance: none;
  }
  
  input[type=range]::-webkit-slider-thumb {
	  border: 2px solid;
	  height: 26px;
	  width: 12px;
	  border-radius: 22px;
	  background: white;
	  cursor: pointer;
	  -webkit-appearance: none;
	  margin-top: -12px;
	  border-color: #029383;
	}
	input[type=range]:hover::-webkit-slider-thumb{
	  border-color: #FBC457;
	}
	input[type=range]::-webkit-slider-runnable-track {
	  height: 2px;
	}
	
	.yellow::-webkit-slider-thumb {
	  border-color: #FBC457 !important;
	}
`;
