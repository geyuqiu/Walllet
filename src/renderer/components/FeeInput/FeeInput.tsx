import React, { useState } from "react";

export const FeeInput = ({ onChange }: Partial<FeeInputProps>) => {
	const [fee, feeChange] = useState("0");
	const handleSliderChange = (event: React.ChangeEvent<HTMLInputElement>) =>
		feeChange(event.target.value);

	const hundredMillion = Math.pow(10, 8);

	const handleTextBoxChange = (
		event: React.ChangeEvent<HTMLInputElement>
	) => {
		const value = parse(event.target.value);
		feeChange(value);
		emitValueToSatoshi(value);
	};

	const emitValueToSatoshi = (value: string): void => {
		if (onChange) {
			onChange(String(Number(value) * hundredMillion));
		}
	};

	const parse = (value: string): string => {
		let replaced = value.replace(/,/g, ".");

		replaced = replaceGreaterThan(replaced, /\./g, 2, "");
		replaced = String(
			Math.round(Number(replaced) * hundredMillion) / hundredMillion
		);
		return replaced;
	};

	const replaceGreaterThan = (
		s: string,
		regex: RegExp,
		index: number,
		replacedWith: string
	): string => {
		let i = 0;
		return s.replace(regex, (match: string) => {
			i += 1;
			if (i >= index) return replacedWith;
			return match;
		});
	};

	return (
		<>
			<label htmlFor="fee">Fee:</label>
			<input
				role="textbox"
				value={fee}
				placeholder="Simple Text"
				onChange={handleTextBoxChange}
			/>
			<input
				type="range"
				id="fee"
				name="fee"
				min="0"
				step="0.00000001"
				role="slider"
				onChange={handleSliderChange}
				value={fee}
			/>
		</>
	);
};

type FeeInputProps = {
	onChange: Function;
};
