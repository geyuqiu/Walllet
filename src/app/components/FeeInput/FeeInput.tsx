import React, {useState} from "react";
import {styled} from "twin.macro";

import {hundredMillion, parse} from './FeeParser';
import {sliderStyle} from './Slider.styles';

const SliderWrapper = styled.div`
	${sliderStyle}
`;

export const FeeInput = ({onChange}: Partial<FeeInputProps>) => {
	const [fee, feeChange] = useState("0");
	const handleSliderChange = (event: React.ChangeEvent<HTMLInputElement>) =>
		feeChange(event.target.value);

	const handleTextBoxChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const value = parse(event.target.value);
		feeChange(value);
		emitValueToSatoshi(value);
	};

	const emitValueToSatoshi = (value: string): void => {
		if (onChange) {
			onChange(String(Number(value) * hundredMillion));
		}
	};

	return (
		<>
			<label htmlFor="fee">Fee:</label>
			<input role="textbox" placeholder="Simple Text"
				onChange={handleTextBoxChange} value={fee}
			/>
			<SliderWrapper>
				<input type="range" id="fee" name="fee" min="0" max="5" step="0.00000001" role="slider"
					onChange={handleSliderChange}
					value={fee}
				/>
			</SliderWrapper>
		</>
	);
};

type FeeInputProps = {
	onChange: Function;
};
