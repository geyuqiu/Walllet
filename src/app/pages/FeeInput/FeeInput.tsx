import React, {useRef, useState} from "react";
import {styled} from "twin.macro";

import {BackButton} from '../../components/BackButton/BackButton';
import {hundredMillion, parse} from './FeeParser';
import {inputRangeStyle} from './InputRangeStyle';

const SliderWrapper = styled.div`
	${inputRangeStyle}
`;
const initialValue = '0';

export const FeeInput = ({onChange}: Partial<FeeInputProps>) => {
	const [sliderValue, setSliderValue] = useState(initialValue);
	const [textBoxValue, setTextBoxValue] = useState('currency: DARK');
	const onInputRangeChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const value = event.target.value;
		setSliderValue(value);
		setTextBoxValue(value);
		setTextBoxStyles();
		setInputRangeColor('#FBC457', Number(value));
	}

	const handleTextBoxChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const value = parse(event.target.value);
		setTextBoxValue(value);
		setTextBoxStyles();
		if (!value.endsWith('.')) {
			const nullSafe = !value ? initialValue : value;
			setSliderValue(nullSafe);
			setInputRangeColor('#046E62', Number(value));
			emitValueToSatoshi(nullSafe);
		}
	};

	const emitValueToSatoshi = (value: string): void => {
		if (onChange) {
			onChange(String(Number(value) * hundredMillion));
		}
	};
	const inputRangeRef = useRef<any>(null);
	const textBoxRef = useRef<any>(null);
	const labelRef = useRef<any>(null);

	const setInputRangeColor = (colorBeforeThumb: string, value: number) => {
		const inputRange = inputRangeRef.current;
		const newValue = (value - inputRange.min) / (inputRange.max - inputRange.min) * 100;
		inputRange.style.background = `linear-gradient(to right, ${colorBeforeThumb} 0%, ${colorBeforeThumb} ${newValue}%, #C7C9CD ${newValue}%, #C7C9CD 100%)`
	};

	const setTextBoxStyles = () => {
		textBoxRef.current.style.color = '#1F2121';
		textBoxRef.current.style.fontWeight = '600';
	}

	const onTextBoxHover = () => labelRef.current.style.color = '#046E62';

	const exitTextBoxHover = () => labelRef.current.style.color = '#1F2121';

	return (
		<section className="p-12 w-1/2">
			<BackButton className="mb-4"/>
			<label className="text-black-dark pl-6 leading-relaxed font-semibold"
			       ref={labelRef}>Fee</label>
			<section onMouseEnter={onTextBoxHover} onMouseLeave={exitTextBoxHover}>
				<input role="textbox" placeholder="currency: DARK"
	        className="p-4 text-gray-darkest hover:border-green-darkest border-gray-darkest border border-solid rounded-50vh w-full focus:outline-none"
					onChange={handleTextBoxChange} value={textBoxValue}
	        ref={textBoxRef}
				/>
				<section className="px-6">
					<SliderWrapper>
						<input type="range" id="fee" name="fee" min="0" max="5" step="0.00000001" role="slider"
							onChange={onInputRangeChange} value={sliderValue}
						  ref={inputRangeRef}
						/>
					</SliderWrapper>
				</section>
			</section>
		</section>
	);
};

type FeeInputProps = {
	onChange: Function;
};
