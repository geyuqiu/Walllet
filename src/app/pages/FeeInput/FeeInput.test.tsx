import {fireEvent, render, screen} from "@testing-library/react";
import React from "react";

import {FeeInput} from "./FeeInput";

describe("FeeInput should", () => {
	it("match snapshot", () => {
		const { container } = render(<FeeInput />);
		expect(container).toMatchSnapshot();
	});

	it("update input when editing slide", () => {
		render(<FeeInput />);

		fireEvent.change(screen.getByRole("slider"), {
			target: {
				value: "0.05678912",
			},
		});

		expect(screen.getByRole("textbox")).toHaveValue("0.05678912");
	});

	it("update slide when editing input", () => {
		render(<FeeInput />);

		fireEvent.change(screen.getByRole("textbox"), {
			target: {
				value: "0.05678912",
			},
		});

		expect(screen.getByRole("slider")).toHaveValue("0.05678912");
	});

	it("emit value in satoshi on input change", () => {
		const onChange = jest.fn();

		render(<FeeInput onChange={onChange} />);

		fireEvent.change(screen.getByRole("textbox"), {
			target: {
				value: "0.123",
			},
		});

		expect(onChange).toHaveBeenCalledWith("12300000");
	});

	it("not show scientific notation", () => {
		render(<FeeInput />);

		fireEvent.change(screen.getByRole("slider"), {
			target: {
				value: "0.000000001",
			},
		});

		expect(screen.getByRole("textbox")).not.toHaveValue("1e-8");
	});

	it("not allow non-numeric characters", () => {
		render(<FeeInput />);

		fireEvent.input(screen.getByRole("textbox"), {
			target: {
				value: "abc012@#9", // go against the design 0.03974945 DARK when sliding the inputRange
			},
		});

		expect(screen.getByRole("textbox")).toHaveValue("0129");
	});

	it("not allow more than 8 decimals (rounds the number)", () => {
		render(<FeeInput />);

		fireEvent.input(screen.getByRole("textbox"), {
			target: {
				value: "0.00000044",
			},
		});

		expect(screen.getByRole("textbox")).toHaveValue("0.00000044");

		fireEvent.input(screen.getByRole("textbox"), {
			target: {
				value: "0.000000447",
			},
		});

		expect(screen.getByRole("textbox")).toHaveValue("0.00000045");

		fireEvent.input(screen.getByRole("textbox"), {
			target: {
				value: "0.0000000123",
			},
		});

		expect(screen.getByRole("textbox")).toHaveValue("0.00000001");

		fireEvent.input(screen.getByRole("textbox"), {
			target: {
				value: "12345678.12",
			},
		});

		expect(screen.getByRole("textbox")).toHaveValue("12345678.12");
	});

	it("not allow more than 1 separator", () => {
		render(<FeeInput />);

		fireEvent.input(screen.getByRole("textbox"), {
			target: {
				value: "1.234.56",
			},
		});

		expect(screen.getByRole("textbox")).toHaveValue("1.23456");

		fireEvent.input(screen.getByRole("textbox"), {
			target: {
				value: "123456",
			},
		});

		expect(screen.getByRole("textbox")).toHaveValue("123456");

		fireEvent.input(screen.getByRole("textbox"), {
			target: {
				value: "1.23.4.5.6",
			},
		});

		expect(screen.getByRole("textbox")).toHaveValue("1.23456");
	});

	it("take the first separator", () => {
		render(<FeeInput />);

		fireEvent.input(screen.getByRole("textbox"), {
			target: {
				value: "1,234.56",
			},
		});

		expect(screen.getByRole("textbox")).toHaveValue("1.23456");

		fireEvent.input(screen.getByRole("textbox"), {
			target: {
				value: "1.234,56",
			},
		});

		expect(screen.getByRole("textbox")).toHaveValue("1.23456");
	});
});
