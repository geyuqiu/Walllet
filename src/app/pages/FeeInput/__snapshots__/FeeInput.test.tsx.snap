// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeeInput should match snapshot 1`] = `
<div>
  <section
    class="p-12 w-1/2"
  >
    <button
      class="BackButton__StyledBackButton-tw1zv6-0 eIXjtA mb-4"
    >
      <div
        class="sc-bdfBwQ iyevwF mx-auto"
        height="20"
        width="20"
      >
        <svg>
          arrow-left.svg
        </svg>
      </div>
    </button>
    <label
      class="text-black-dark pl-6 leading-relaxed font-semibold"
    >
      Fee
    </label>
    <section>
      <input
        class="p-4 text-gray-darkest hover:border-green-darkest border-gray-darkest border border-solid rounded-50vh w-full focus:outline-none"
        placeholder="currency: DARK"
        role="textbox"
        value="currency: DARK"
      />
      <section
        class="px-6"
      >
        <div
          class="FeeInput__SliderWrapper-sc-1tci9nq-0 kMhBBq"
        >
          <input
            id="fee"
            max="5"
            min="0"
            name="fee"
            role="slider"
            step="0.00000001"
            type="range"
            value="0"
          />
        </div>
      </section>
    </section>
  </section>
</div>
`;
